### PR TITLE
feat: refer datavolumes

### PR DESCRIPTION
**Pull request by thxCode**
#101

- Introduce CDI v1beta1 APIs and update the dependencies
- Ref in-use DataVolumes (add an annotation value refer to related VirtualMachine).

### CODE REVIEW

File: pkg/controller/master/virtualmachine/vmi_controller.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+			// which triggers the cascading deletion of the VirtualMachineInstance.
+			// however, the deletion of the VirtualMachineInstance is placed in the background,
+			// the VirtualMachineInstance is still deleting, but the corresponding VirtualMachine has been deleted.
+			// so we can ignores this error.
+			return vmi, nil
+		}
+		return vmi, fmt.Errorf("failed to get VirtualMachine referred by VirtualMachineInstance(%s/%s): %w", vmi.Namespace, vmi.Name, err)
+	}
+	if vm.DeletionTimestamp != nil {
+		// doesn't process as there is another unref reconciling executed by VirtualMachine controller
```
Comment (guangbochen) [inline pkg/controller/master/virtualmachine/vmi_controller.go]: gramma: reconciling executed by the VirtualMachine controller and don't process in here.

File: pkg/controller/master/virtualmachine/vmi_controller.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+		return vmi, nil
+	}
+
+	// collects all desired DataVolumes from VirtualMachine
+	var dataVolumeNamesDesired = sets.String{}
```
Comment (guangbochen) [inline pkg/controller/master/virtualmachine/vmi_controller.go]: prefer to call `dataVolumeNames` as the comment already explains it.

File: pkg/controller/master/virtualmachine/vmi_controller.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+	var dataVolumeNameObserved = sets.String{}
+	for _, volume := range vmi.Spec.Volumes {
+		if volume.DataVolume != nil && volume.DataVolume.Name != "" {
+			dataVolumeNameObserved.Insert(volume.DataVolume.Name)
+		}
+	}
+
+	// abandons useless DataVolumes
+	var dataVolumeNamespace = vmi.Namespace
+	var uselessDataVolumeNames = dataVolumeNameObserved.Difference(dataVolumeNamesDesired).List()
```
Comment (guangbochen) [inline pkg/controller/master/virtualmachine/vmi_controller.go]: ditto `ownerlessDataVolumeNames `

File: pkg/controller/master/virtualmachine/vmi_controller.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+	// collects all observed in-use DataVolumes from VirtualMachineInstance
+	var dataVolumeNameObserved = sets.String{}
+	for _, volume := range vmi.Spec.Volumes {
+		if volume.DataVolume != nil && volume.DataVolume.Name != "" {
+			dataVolumeNameObserved.Insert(volume.DataVolume.Name)
+		}
+	}
+
+	// abandons useless DataVolumes
```
Comment (guangbochen) [inline pkg/controller/master/virtualmachine/vmi_controller.go]: use `ownerless`

File: pkg/controller/master/virtualmachine/datavolume_util.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+		return err
+	}
+	if !orphan {
+		return nil
+	}
+	_, err = dataVolumeClient.Update(dvHandler.(*cdiapis.DataVolume))
+	return err
+}
+
+// abandonDataVolumeIfOwned tries to un-refer the target VirtualMachine as the owner of the DataVolume.
```
Comment (guangbochen) [inline pkg/controller/master/virtualmachine/datavolume_util.go]: un-refer is not a word, change to `unset the dataVolume's owner reference of its target VM`.

File: pkg/ref/owner.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+type MetadataRuntimeObject interface {
+	metav1.Object
+	runtime.Object
+}
+
+// RecordOwner adds an non-controller OwnerReference of object to refer the given owner,
+// returns a "false" signal to indicate that the owner was a reference.
+//   It's worth to notice that the input object must be a runtime root object of Kubernetes API, otherwise,
+//   an error will be received.
+func RecordOwner(obj MetadataRuntimeObject, owner MetadataRuntimeObject, ownerGVK schema.GroupVersionKind) (MetadataRuntimeObject, bool, error) {
```
Comment (guangbochen) [inline pkg/ref/owner.go]: prefer to use `GrantOwner`

File: pkg/controller/master/virtualmachine/datavolume_util.go [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+func abandonDataVolumeIfOwned(dataVolumeClient cdictrl.DataVolumeClient, dv *cdiapis.DataVolume, vm *kubevirtapis.VirtualMachine) error {
+	if vm == nil || dv == nil || dv.DeletionTimestamp != nil {
+		return nil
+	}
+	var dvHandler, hasOwned, err = ref.EraseOwner(dv, vm, kubevirtapis.VirtualMachineGroupVersionKind)
```
Comment (guangbochen) [inline pkg/controller/master/virtualmachine/datavolume_util.go]: either use hasOwner or isOwned

File: pkg/controller/master/virtualmachine/vm_controller.go [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+	var outTreeDataVolumeNames = getOutTreeDataVolumeNames(&vm.Spec)
+
+	// adopts orphan out-tree DataVolumes
+	var dataVolumeNamespace = vm.Namespace
+	for _, dataVolumeName := range outTreeDataVolumeNames {
+		var dataVolume, err = h.dataVolumeCache.Get(dataVolumeNamespace, dataVolumeName)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				// NB(thxCode): ignores error, since this can't be fixed by an immediate requeue,
```
Comment (guangbochen) [inline pkg/controller/master/virtualmachine/vm_controller.go:31]: requeue -> `re-sync/re-queue`

File: pkg/controller/master/virtualmachine/vm_controller.go [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+
+// RefOutTreeDataVolumes records the target VirtualMachine as the owner of the out-tree DataVolumes.
+func (h *VMController) RefOutTreeDataVolumes(_ string, vm *kubevirtapis.VirtualMachine) (*kubevirtapis.VirtualMachine, error) {
+	if vm == nil || vm.DeletionTimestamp != nil || vm.Spec.Template == nil {
+		return vm, nil
+	}
+
+	var outTreeDataVolumeNames = getOutTreeDataVolumeNames(&vm.Spec)
+
+	// adopts orphan out-tree DataVolumes
```
Comment (guangbochen) [inline pkg/controller/master/virtualmachine/vm_controller.go]: use ownerless

File: pkg/controller/master/virtualmachine/vm_controller.go [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+)
+
+type VMController struct {
+	// dv ctrl suite
+	dataVolumeClient cdictrl.DataVolumeClient
+	dataVolumeCache  cdictrl.DataVolumeCache
+}
+
+// RefOutTreeDataVolumes records the target VirtualMachine as the owner of the out-tree DataVolumes.
+func (h *VMController) RefOutTreeDataVolumes(_ string, vm *kubevirtapis.VirtualMachine) (*kubevirtapis.VirtualMachine, error) {
```
Comment (guangbochen) [inline pkg/controller/master/virtualmachine/vm_controller.go]: prefer to call it `SetOwnerOfOutTreeDataVolumes`

File: pkg/controller/master/virtualmachine/vm_controller.go [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+			return vm, fmt.Errorf("failed to ref VitrualMachine(%s/%s) as DataVolume(%s/%s)'s owner: %w",
+				vm.Namespace, vm.Name, dataVolumeNamespace, dataVolumeName, err)
+		}
+	}
+
+	return vm, nil
+}
+
+// RefOutTreeDataVolumes erases the target VirtualMachine from the owner of the out-tree DataVolumes.
+func (h *VMController) UnrefOutTreeDataVolumes(_ string, vm *kubevirtapis.VirtualMachine) (*kubevirtapis.VirtualMachine, error) {
```
Comment (guangbochen) [inline pkg/controller/master/virtualmachine/vm_controller.go]: prefer to call it `RemoveOwnerOfOutTreeDataVolumes`

File: go.mod [hunk 12]
```diff
@@ -10,6 +10,7 @@ replace (
 	github.com/rancher/steve => github.com/rancheredge/steve v0.0.0-20200708031911-f69e0f4820b4
 	k8s.io/client-go => k8s.io/client-go v0.18.0
 	kubevirt.io/client-go => github.com/orangedeng/client-go v0.31.1-0.20200715061104-844cb60487e4
+	kubevirt.io/containerized-data-importer => github.com/thxcode/kubevirt-containerized-data-importer v1.22.0-apis-only
```
[the remark this answers is not in the document]
Comment (thxCode) [reply go.mod:13]: OK, perhaps we should move `kubevirt.io/client-go` and `kubevirt.io/containered-data-importer` to this repo latter. I prefer to use another PR to change them.

File: pkg/controller/master/virtualmachine/vm_controller.go [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	kubevirtapis "kubevirt.io/client-go/api/v1alpha3"
+
+	cdictrl "github.com/rancher/harvester/pkg/generated/controllers/cdi.kubevirt.io/v1beta1"
+)
+
+type VMController struct {
+	// dv ctrl suite
```
[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/controller/master/virtualmachine/vm_controller.go]: Removed

File: pkg/controller/master/virtualmachine/register.go [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+	"context"
+
+	"github.com/rancher/harvester/pkg/config"
+)
+
+func Register(ctx context.Context, management *config.Management) error {
+	var dataVolumeClient = management.CDIFactory.Cdi().V1beta1().DataVolume()
+	var dataVolumeCache = dataVolumeClient.Cache()
+
+	// registers vm controller
```
[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/controller/master/virtualmachine/register.go]: Updated

File: pkg/config/context.go [hunk 15]
```diff
@@ -99,12 +110,19 @@ func setupManagement(ctx context.Context, restConfig *rest.Config, opts *generic
...
 
-	harvester, err := harvester.NewFactoryFromConfigWithOptions(restConfig, opts)
+	cdiFactory, err := cdi.NewFactoryFromConfigWithOptions(restConfig, opts)
+	if err != nil {
+		return nil, err
+	}
+	management.CDIFactory = cdiFactory
+	management.starters = append(management.starters, cdiFactory)
+
+	harv, err := harvester.NewFactoryFromConfigWithOptions(restConfig, opts)
```
[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/config/context.go:120]: Updated

File: pkg/controller/master/virtualmachine/vmi_controller.go [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+	cdictrl "github.com/rancher/harvester/pkg/generated/controllers/cdi.kubevirt.io/v1beta1"
+	kubevirtctrl "github.com/rancher/harvester/pkg/generated/controllers/kubevirt.io/v1alpha3"
+	"github.com/rancher/harvester/pkg/ref"
+)
+
+type VMIController struct {
+	// vm ctrl suite
+	virtualMachineCache kubevirtctrl.VirtualMachineCache
+
+	// dv ctrl suite
```
[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/controller/master/virtualmachine/vmi_controller.go]: Removed

File: pkg/ref/owner.go [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/utils/pointer"
+)
+
+// GetControllerOfNoCopy returns a pointer to the controllerRef if controllee has a controller.
+func GetControllerOfNoCopy(controllee metav1.Object) *metav1.OwnerReference {
```
[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/ref/owner.go]: Updated, `controllee` means the object is controlled by the `controller`, this word is copied from https://github.com/kubernetes/apimachinery/blob/d8e5c2b33a59167f06f5fdca3973d0ea179f9a7c/pkg/apis/meta/v1/controller_ref.go#L32-L40.

File: pkg/ref/owner.go [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+// MetadataRuntimeObject combines the interfaces of Kubernetes API metadata and Runtime Object.
+type MetadataRuntimeObject interface {
+	metav1.Object
+	runtime.Object
+}
+
+// RecordOwner adds an non-controller OwnerReference of object to refer the given owner,
+// returns a "false" signal to indicate that the owner was a reference.
+//   It's worth to notice that the input object must be a runtime root object of Kubernetes API, otherwise,
+//   an error will be received.
```
[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/ref/owner.go]: Updated

File: pkg/ref/owner.go [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+		return obj, false, fmt.Errorf("object %T is not a runtime root object of Kubernetes API", obj)
+	}
+	objCp.SetOwnerReferences(refs)
+	return objCp, true, nil
+}
+
+// EraseOwner removes the non-controller OwnerReference of object to refer the given owner,
+// returns a "false" signal to indicate that the owner isn't a reference.
+//   It's worth to notice that the input object must be a runtime root object of Kubernetes API, otherwise,
+//   an error will be received.
```
[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/ref/owner.go]: Updated

File: pkg/ref/owner.go [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+func RecordOwner(obj MetadataRuntimeObject, owner MetadataRuntimeObject, ownerGVK schema.GroupVersionKind) (MetadataRuntimeObject, bool, error) {
+	if IsOwnedBy(obj, owner) {
+		return obj, false, nil
+	}
+
+	var refs = obj.GetOwnerReferences()
+	var refPtr = NewOwnerRef(owner, ownerGVK)
+	refs = append(refs, *refPtr)
+
+	var objCp, ok = obj.DeepCopyObject().(MetadataRuntimeObject)
```
[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/ref/owner.go]: Updated

File: pkg/ref/owner.go [hunk 21]
```diff
@@ -0,0 +1,95 @@
...
+	for _, ref := range obj.GetOwnerReferences() {
+		if ref.UID == ownerUID &&
+			ref.APIVersion == ownerAPIVersion &&
+			ref.Kind == ownerKind {
+			continue
+		}
+		refs = append(refs, ref)
+	}
+
+	var objCp, ok = obj.DeepCopyObject().(MetadataRuntimeObject)
```
[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/ref/owner.go]: Updated

File: pkg/controller/master/virtualmachine/vmi_controller.go [hunk 22]
```diff
@@ -0,0 +1,141 @@
...
+//   -           name: data-disk-2
+//           volumes:
+//           - dataVolume:
+//             name: root-disk
+//           - dataVolume:
+//             name: data-disk-1
+//   -       - dataVolume:
+//   -         name: data-disk-2
+//   ```
+func (h *VMIController) UnrefUselessDataVolumes(_ string, vmi *kubevirtapis.VirtualMachineInstance) (*kubevirtapis.VirtualMachineInstance, error) {
```
[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/controller/master/virtualmachine/vmi_controller.go]: Updated

File: pkg/controller/master/virtualmachine/vmi_controller.go [hunk 23]
```diff
@@ -0,0 +1,141 @@
...
+	if vmi == nil || vmi.DeletionTimestamp == nil {
+		return vmi, nil
+	}
+
+	// NB(thxCode): if the VirtualMachineInstance is generated by VirtualMachine,
+	// it must have an OwnerReference ref to a VirtualMachine,
+	// so we can validate this reference and process this kind of VirtualMachineInstance.
```
[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/controller/master/virtualmachine/vmi_controller.go]: Updated

File: pkg/controller/master/virtualmachine/vmi_controller.go [hunk 24]
```diff
@@ -0,0 +1,141 @@
...
+	var vmReferred = ref.GetControllerOfNoCopy(vmi)
+	if vmReferred == nil {
+		// doesn't process headless VirtualMachineInstance
+		return vmi, nil
+	}
+	var vmGVK = kubevirtapis.VirtualMachineGroupVersionKind
+	if vmReferred.APIVersion != vmGVK.GroupVersion().String() ||
+		vmReferred.Kind != vmGVK.Kind {
+		// doesn't process none VirtualMachine type managed VirtualMachineInstance
```
[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/controller/master/virtualmachine/vmi_controller.go:80]: Updated

File: pkg/controller/master/virtualmachine/vmi_controller.go [hunk 25]
```diff
@@ -0,0 +1,141 @@
...
+	if vmi == nil || vmi.DeletionTimestamp == nil {
+		return vmi, nil
+	}
+
+	// NB(thxCode): if the VirtualMachineInstance is generated by VirtualMachine,
+	// it must have an OwnerReference ref to a VirtualMachine,
+	// so we can validate this reference and process this kind of VirtualMachineInstance.
+	var vmReferred = ref.GetControllerOfNoCopy(vmi)
+	if vmReferred == nil {
+		// doesn't process headless VirtualMachineInstance
```
[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/controller/master/virtualmachine/vmi_controller.go]: Updated

File: pkg/controller/master/virtualmachine/vmi_controller.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/controller/master/virtualmachine/vmi_controller.go]: Updated

File: pkg/controller/master/virtualmachine/vmi_controller.go (hunk 3 above)
Comment (thxCode) [reply pkg/controller/master/virtualmachine/vmi_controller.go]: Updated

File: pkg/controller/master/virtualmachine/vmi_controller.go (hunk 1 above)
Comment (thxCode) [reply pkg/controller/master/virtualmachine/vmi_controller.go]: Updated

File: pkg/controller/master/virtualmachine/vmi_controller.go (hunk 2 above)
Comment (thxCode) [reply pkg/controller/master/virtualmachine/vmi_controller.go]: Updated

File: pkg/controller/master/virtualmachine/datavolume_util.go (hunk 5 above)
Comment (thxCode) [reply pkg/controller/master/virtualmachine/datavolume_util.go]: Updated

File: pkg/controller/master/virtualmachine/vmi_controller.go (hunk 4 above)
Comment (thxCode) [reply pkg/controller/master/virtualmachine/vmi_controller.go]: Updated

File: pkg/controller/master/virtualmachine/datavolume_util.go (hunk 7 above)
Comment (thxCode) [reply pkg/controller/master/virtualmachine/datavolume_util.go]: Updated

File: pkg/controller/master/virtualmachine/vm_controller.go (hunk 9 above)
Comment (thxCode) [reply pkg/controller/master/virtualmachine/vm_controller.go]: Removed this comment

File: pkg/controller/master/virtualmachine/vm_controller.go (hunk 10 above)
Comment (thxCode) [reply pkg/controller/master/virtualmachine/vm_controller.go]: Updated

File: pkg/controller/master/virtualmachine/vm_controller.go (hunk 11 above)
Comment (thxCode) [reply pkg/controller/master/virtualmachine/vm_controller.go]: Changed to `UnsetOwnerOfOutTreeDataVolumes`

File: pkg/controller/master/virtualmachine/vm_controller.go (hunk 8 above)
Comment (thxCode) [reply pkg/controller/master/virtualmachine/vm_controller.go:31]: I think `requeue` is enough: https://github.com/kubernetes-sigs/controller-runtime/search?q=requeue&unscoped_q=requeue.

File: pkg/ref/owner.go (hunk 6 above)
Comment (thxCode) [reply pkg/ref/owner.go]: Updated

File: pkg/controller/master/virtualmachine/register.go [hunk 26]
```diff
@@ -0,0 +1,33 @@
...
+	var vmCtrl = &VMController{
+		dataVolumeClient: dataVolumeClient,
+		dataVolumeCache:  dataVolumeCache,
+	}
+	var virtualMachineClient = management.VirtFactory.Kubevirt().V1alpha3().VirtualMachine()
+	virtualMachineClient.OnChange(ctx, "VMController.RefOutTreeDataVolumes", vmCtrl.RefOutTreeDataVolumes)
+	virtualMachineClient.OnRemove(ctx, "VMController.UnrefOutTreeDataVolumes", vmCtrl.UnrefOutTreeDataVolumes)
```
[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/controller/master/virtualmachine/register.go]: I don't know why we should `constant` these keys, there are not any other references for them. Although the keys of `OnRemove` are used as a part of the finalizer, these `OnRemove` keys are useless for other logic.

File: go.mod (hunk 12 above)
[the remark this answers is not in the document]
Comment (guangbochen) [reply go.mod:13]: sure, let's manage this change in a separate PR, thanks,

File: pkg/controller/master/virtualmachine/register.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (guangbochen) [reply pkg/controller/master/virtualmachine/register.go]: this is just to keep consistent with other controller registers, let's make sure the name is updated if u prefer to keep within the method, thanks.

[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/controller/master/virtualmachine/register.go]: Updated

File: pkg/controller/master/virtualmachine/register.go [hunk 27]
```diff
@@ -0,0 +1,33 @@
...
+
+	// registers vmi controller
+	var virtualMachineCache = virtualMachineClient.Cache()
+	var vmiCtrl = &VMIController{
+		virtualMachineCache: virtualMachineCache,
+		dataVolumeClient:    dataVolumeClient,
+		dataVolumeCache:     dataVolumeCache,
+	}
+	var virtualMachineInstanceClient = management.VirtFactory.Kubevirt().V1alpha3().VirtualMachineInstance()
+	virtualMachineInstanceClient.OnRemove(ctx, "VMIController.UnrefUselessDataVolumes", vmiCtrl.UnrefUselessDataVolumes)
```
[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/controller/master/virtualmachine/register.go]: Updated

File: pkg/controller/master/virtualmachine/vm_controller.go [hunk 28]
```diff
@@ -0,0 +1,97 @@
...
+				// NB(thxCode): ignores error, since this can't be fixed by an immediate requeue,
+				// and also doesn't block the whole logic if the DataVolume has already been deleted.
+				continue
+			}
+			return vm, fmt.Errorf("failed to get DataVolume(%s/%s): %w", dataVolumeNamespace, dataVolumeName, err)
+		}
+
+		err = unsetBoundedDataVolumeReference(h.dataVolumeClient, dataVolume, vm)
+		if err != nil {
+			return vm, fmt.Errorf("failed to unref VitrualMachine(%s/%s) as DataVolume(%s/%s)'s owner: %w",
```
[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/controller/master/virtualmachine/vm_controller.go]: Updated

File: pkg/controller/master/virtualmachine/vm_controller.go [hunk 29]
```diff
@@ -0,0 +1,97 @@
...
+				// NB(thxCode): ignores error, since this can't be fixed by an immediate requeue,
+				// when the DataVolume is ready, the VirtualMachine resource will requeue.
+				continue
+			}
+			return vm, fmt.Errorf("failed to get DataVolume(%s/%s): %w", dataVolumeNamespace, dataVolumeName, err)
+		}
+
+		err = setOwnerlessDataVolumeReference(h.dataVolumeClient, dataVolume, vm)
+		if err != nil {
+			return vm, fmt.Errorf("failed to ref VitrualMachine(%s/%s) as DataVolume(%s/%s)'s owner: %w",
```
[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/controller/master/virtualmachine/vm_controller.go]: Updated

**Review by guangbochen — APPROVED** [review]

**Comment by gitlawr** [comment]
If I understand correctly owner references are used for [garbage collection](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/) in Kubernetes
We are adding owner references on VM creation and removing them on deletion while keeping the owned DV objects.
It looks different from what the issue claims. What's the motivation for the change?

**Comment by thxCode** [comment]
~@gitlawr, I have sorted out why I change to use owner reference as below:~

~**What are we facing:**~
- ~Target `VirtualMachine` doesn't refer to the out-tree `DataVolume`. ~
- ~When removing the definition of `DataVolume` from `dataVolumeTemplates`, those removed ones still refer to the target `VirtualMachine`.~

~**What can we find from above:**~
1. ~There is no place to describe the relationship between the two parties involved.~
1. ~A descriptive relationship is maintained by the two parties who have no relationship.~

~**What are we expected:**~
- ~Referring to the out-tree `DataVolume` when the target `VirtualMachine` is using it.~
- ~Unreferring the `DataVolume` when the target `VirtualMachine` is not using it.~
> ~This rule is breaking the above conclusion 2, the user needs to confirm the `DataVolume` if want to delete.~

~**Why not use annotation/label:**~
1. ~The annotation is used to describe the metadata of that resource, the relationship is not metadata, it is a connection.~
1. ~The label is good to describe the relationship, but it is too discrete.~
1. ~The annotation and label are not good at describing multiple, I mean either comma-separated value or key-prefix map is not an intuitive experience.~

~**Why use owner reference:**~
1. ~For out-tree `DataVolume`, we can keep the same behavior as what KubeVirt did for referred `DataVolume`.~
1. ~Unify the cascade deletion experience, when using `kubectl`, the user can determine to delete the out-tree `DataVolume` or not.~
1. ~The owner reference is maintained by Kubernetes, when the harvester is crashing, a `DataVolume` which has been referred multiple don't keep the stale relationships.~

~**Notice:**~
1. ~Neither using owner reference nor using annotation/label is a forbidden action, which means the user can delete the `DataVolume` during it is referring.~
1. ~In harvester, either using owner reference or using annotation/label needs to append a finalizer to `VirtualMachine` and `VirtualMachineInstance` to unrefer the `DataVolume` when they are deleting.~

~In short, the benefits of using owner reference in engineering logic than using annotaion/label are:~
1. ~The `DataVolume` that has been referred many times is well protected. We can even refer a `DataVolume` created by another VirtualMachine`dataVolumeTemplates` in the new `VirtualMachine`, although this is not recommend.~
1. ~Easier to ensure the correctness of the reference relationship between `DataVolume` and `VirtualMachine`. We can still delete `VirtualMachine` via `kubectl` during harvester isn't work, which means we don't need to modify the `DataVolume` annotation/label manually.~
1. ~Programming in owner reference is easier than operating the annotation/label.~

**Comment by gitlawr** [comment]
I see. Will we change the original kubevirt behaviors? When a VM is removed, DVs defined in DVT are removed. DVs defined externally are not.

**Comment by thxCode** [comment]
> **What can we find from above:**
> 1. There is no place to describe the relationship between the two parties involved.
> 1. A descriptive relationship is maintained by the two parties who have no relationship.
> 
> **What are we expected:**
> - Referring to the out-tree DataVolume when the target VirtualMachine is using it.
> - Unreferring the DataVolume when the target VirtualMachine is not using it.
> > This rule is breaking the above conclusion 2, the user needs to confirm the DataVolume if want to delete

As the above said, this PR will do:

1. For out-tree `DataVolume`, if a `VirtualMachine` wants to use it, it will be referred util the `VirtualMachine` is deleted or doesn't use that out-tree `DataVolume` anymore. 
    > _So the out-tree `DataVolume` keep the same behaviour as before._
1. For template `DataVolume`, if the parent `VirtualMachine` removed it, it will be un-referred from the parent `VirtualMachine`. 
    > _So the removed template `DataVolume` will not be removed when deleting the parent `VirtualMachine`, we need to remove it manually._

`#2` doesn't work as KubeVirt.

**Comment by thxCode** [comment]
According to the discussion with @gitlawr and @aiwantaozi offline, we think that it's better to respect the behavior of KubeVirt, so I will change this back to use annotations to indicate the relationship between `DataVolume` and `VirtualMachine`. cc @guangbochen .

**Comment by thxCode** [comment]
Changed, @gitlawr. If this is approved, the frontend should change the logic of finding the owner of the `DataVolume`, @WuJun2016 . 

The value of `harvester.cattle.io/owned-by` is a JSON array string, where the `schema` string filed specifies the steve schema ID of the owner CR and the `refs` array field represents the steve-style data ID(namespaced-scope: `{namespace}/{name}`; cluster-scope: `{name}`) of the owner CRs.

File: pkg/ref/api_id.go [hunk 30]
```diff
@@ -0,0 +1,23 @@
...
+)
+
+// Parse parses the steve api ID.
+func Parse(ref string) (namespace string, name string) {
+	parts := strings.SplitN(ref, ":", 2)
+	if len(parts) == 1 {
+		return "", parts[0]
+	}
+	return parts[0], parts[1]
+}
```
Comment (gitlawr) [inline pkg/ref/api_id.go:15]: Looks like steve API ID is in the format of `default/name`, so we should either change the function description or change the format we are using

File: pkg/ref/annotation.go [hunk 31]
```diff
@@ -0,0 +1,196 @@
...
+
+	var schemaRef, existed = o[schemaID]
+	if !existed {
+		return false
+	}
+	return schemaRef.SchemaID == schemaID && schemaRef.References.Has(ownerRef)
+}
+
+// Add adds the given owner as an annotation schema owner,
+// returns a "false" signal to indicate that the ownerRef was a reference.
```
[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/ref/annotation.go]: Updated

File: pkg/ref/api_id.go (hunk 30 above)
Comment (thxCode) [reply pkg/ref/api_id.go:15]: `Parse` method will change other UI APIs, I prefer to update the `Construct` method in this PR, and use another new PR to repair the `Parse` method, cc @aiwantaozi .

**Review by gitlawr — APPROVED** [review]
LGTM

**Pull request merged**